### PR TITLE
Mesheryctl service type

### LINKED ISSUE #2968: [mesheryctl] Context parameter to control service type deployed
#### Current Behavior
Currently meshery when deployed in kubernetes, is deployed as service type NodePort by default.

#### Desired Behavior
Moving forward, the service type needs to be dynamically configured based on the user's preference, adding a context paramter called `serviceType` would suffice this need.

- Host OS: Mac Linux Windows
- Meshery Version: all
- Platform: Kubernetes

#### Implementation
- Add a context parameter to support different service types.
- Change the service type in the manifest, based on the context parameter value.
- 
---

#### Contributor Resources

- [mesheryctl Contributing Guide](https://github.com/layer5io/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- [Layer5 Community Slack](http://slack.layer5.io)

### PR DESCRIPTION
**Pull request by vineethvanga18**
Signed-off-by: vineethvanga18 <reddy.8@iitj.ac.in>

**Description**
Added t flag for passing service type ex: `./mesheryctl system context create local2 -t ClusterIP` with default as LoadBalancer

This PR fixes #2968 

**Notes for Reviewers**
1. `system context create` creates/adds context to `config.yaml` with a parameter serviceType.
2. When this context is used to `system start` it sets service type in `meshery-service.yaml` and applies this manifest
3.  One issue is that sometimes parameters of `meshery-service.yaml` are being written to config.yaml probably during marshalling.

**[Signed commits](https://github.com/layer5io/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by metonymic-smokey** [comment]
@vineethvanga18 this seems to add service type correctly in meshconfig for me.  
I did not see any changes in `meshery.yaml` though. 
Could you show me an example of the overwriting you mentioned?  

Thanks!

**Comment by vineethvanga18** [comment]
apiversion: v1
contexts:
  local:
    endpoint: http://localhost:30840
    token: Default
    platform: kubernetes
    adapters:
    - meshery-istio
    - meshery-linkerd
    - meshery-consul
    - meshery-nsm
    - meshery-kuma
    - meshery-cpx
    - meshery-osm
    - meshery-traefik-mesh
    channel: stable
    version: latest
    servicetype: LoadBalancer
  local3:
    adapters:
    - meshery-istio
    - meshery-linkerd
    - meshery-consul
    - meshery-nsm
    - meshery-kuma
    - meshery-cpx
    - meshery-osm
    - meshery-traefik-mesh
    channel: stable
    endpoint: http://localhost:9081
    platform: kubernetes
    servicetype: ClusterIP
    token: Default
    version: latest
current-context: local
kind: Service
metadata:
  name: meshery
spec:
  selector:
    io.kompose.service: meshery
  ports:
  - name: http
    port: 9081
    targetPort: 8080
  type: LoadBalancer
tokens:
- location: auth.json
  name: Default

**Comment by vineethvanga18** [comment]
@metonymic-smokey Also changes occur in meshery-service.yaml after you run `mesheryctl system start` using the created context

File: mesheryctl/pkg/utils/platform.go [hunk 1]
```diff
@@ -453,6 +453,49 @@ func ChangeManifestVersion(fileName string, version string, filePath string) err
...
+	ViperK8s.Set("kind", compose.Kind)
+	ViperK8s.Set("metadata", compose.Metadata)
+	ViperK8s.Set("spec", compose.Spec)
+
+	// Marshal the structs
+	newConfig, err := yaml.Marshal(compose)
+	if err != nil {
+		return fmt.Errorf("unable to marshal config %s | %s", fileName, err)
+	}
+	err = ioutil.WriteFile(filePath, newConfig, 0644)
```
Comment (vineethvanga18) [inline mesheryctl/pkg/utils/platform.go:491]: ViperK8s seems to store values for config.yaml and using the same here making it store apiVersion, kind, etc.. and when marshalling this data is also added to config.yaml. Should I create another viper instance to store these or use vipercompose? @navendu-pottekkat

Comment (vineethvanga18) [reply mesheryctl/pkg/utils/platform.go:491]: This was the reason for overwriting @metonymic-smokey

Comment (navendu-pottekkat) [reply mesheryctl/pkg/utils/platform.go:491]: @vineethvanga18 We should have different viper instances for different config files. But I think you are already doing that

Comment (vineethvanga18) [reply mesheryctl/pkg/utils/platform.go:491]: No. ViperK8s is being used for config.yaml. I will create another instance for this

**Review by navendu-pottekkat — COMMENTED** [review]
Do we consider what happens when the user does `system update`? i.e anytime the manifest files are re downloaded?

File: mesheryctl/pkg/utils/platform.go (hunk 1 above)
Comment (navendu-pottekkat) [reply mesheryctl/pkg/utils/platform.go:491]: Oh, yes, I see it now-

Screenshot from 2021-05-26 16-51-19

It is being used to change the endpoints in config.yaml

**Comment by vineethvanga18** [comment]
No. But should be done. We just need to call the above ChangeManifestServiceType function. Should I do that?

**Comment by leecalcote** [comment]
A conflict to attend here - https://github.com/layer5io/meshery/pull/3003/conflicts.

**Comment by navendu-pottekkat** [comment]
This change is needs to be updated with the refactor. Will need to open a new PR. Closing this...

**Pull request closed without merging**